// n8n-nodes-base/credentials/GhostContentApi.credentials.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GhostContentApi = void 0;
class GhostContentApi {
    constructor() {
        this.name = 'ghostContentApi';
        this.displayName = 'Ghost Content API';
        this.documentationUrl = 'ghost';
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                default: '',
                placeholder: 'http://localhost:3001',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/ghost/api/v3/content/settings/',
                method: 'GET',
            },
        };
    }
    authenticate(credentials, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            requestOptions.qs = Object.assign(Object.assign({}, requestOptions.qs), { 'key': credentials.apiKey });
            return requestOptions;
        });
    }
}
exports.GhostContentApi = GhostContentApi;
